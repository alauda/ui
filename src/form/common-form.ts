import {
  ChangeDetectorRef,
  Directive,
  EventEmitter,
  Input,
  Output,
} from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { ReplaySubject } from 'rxjs';

import { coerceAttrBoolean } from '../utils';

/**
 * form control 内部不保持 value 属性，当不使用 ngModel 时可以通过 [(value)] 实现双向绑定；
 * 也可以分别传入 value 和 监听 valueChange 事件以实现阻断用户操作的功能。
 */
@Directive()
// tslint:disable-next-line:directive-class-suffix
export class CommonFormControl<V, M = V> implements ControlValueAccessor {
  @Input()
  get disabled() {
    return this._disabled;
  }

  set disabled(val: boolean | '') {
    this._disabled = coerceAttrBoolean(val);
  }

  @Input()
  get readonly() {
    return this.disabled;
  }

  set readonly(val) {
    this.disabled = val;
  }

  @Input('value')
  get propValue() {
    return this._propValue;
  }

  set propValue(val) {
    if (val === this._propValue) {
      return;
    }
    this._propValue = val;
    this.writeValue(val);
  }

  @Output()
  valueChange = new EventEmitter<V>();

  protected onChange: (_: V) => void;
  protected onTouched: () => void;
  private _propValue: V;
  private _disabled = false;

  model: M = null;
  model$ = new ReplaySubject<M>(1);

  constructor(protected cdr: ChangeDetectorRef) {
    this.model$.subscribe(model => {
      this.model = model;
      this.cdr.markForCheck();
    });
  }

  registerOnChange(fn: (_: V) => void) {
    this.onChange = fn;
  }

  registerOnTouched(fn: () => void) {
    this.onTouched = fn;
  }

  setDisabledState(isDisabled: boolean) {
    this.disabled = isDisabled;
    this.cdr.markForCheck();
  }

  writeValue(value: V) {
    this.model$.next(this.valueIn(value));
  }

  emitModel(model: M) {
    this.emitValue(this.modelOut(model));
  }

  emitValue(value: V) {
    if (this.onChange) {
      this.onChange(value);
      this.writeValue(value);
    }
    this.valueChange.emit(value);
  }

  protected valueIn(value: V): M {
    return value as any;
  }

  protected modelOut(model: M): V {
    return model as any;
  }
}

/**
 * @deprecated use {@link CommonFormControl} instead
 */
export const CommonForm = CommonFormControl;
