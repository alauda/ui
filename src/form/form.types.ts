import { ValueOf } from '../types';

export const LabelPosition = {
  Top: 'top',
  Left: 'left',
  Right: 'right',
} as const;

export enum FormItemWidth {
  Small = 'small',
  Medium = 'medium',
  Large = 'large',
}

export type LabelPosition = ValueOf<typeof LabelPosition>;
