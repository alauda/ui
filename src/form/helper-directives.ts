import { Directive, Input } from '@angular/core';

import { coerceAttrBoolean } from '../utils';

@Directive({
  selector: 'label[auiFormItemLabel]',
  host: {
    '[class.aui-form-item__label]': 'true',
  },
})
export class FormItemLabelDirective {}

@Directive({
  selector: '[auiFormItemHint]',
  host: {
    '[class.aui-form-item__hint]': 'true',
  },
})
export class FormItemHintDirective {}

@Directive({
  selector: '[auiFormItemAddon]',
  host: {
    '[class.aui-form-item__addon]': 'true',
  },
})
export class FormItemAddonDirective {}

@Directive({
  selector: '[auiFormItemError]',
  host: {
    '[class.aui-form-item__error]': 'true',
  },
})
export class FormItemErrorDirective {}

@Directive({
  selector: '[auiFormItemControl]',
  host: {
    '[class.aui-form-item__control]': 'true',
  },
})
export class FormItemControlDirective {
  @Input()
  get required() {
    return this._required;
  }

  set required(val: boolean | '') {
    this._required = coerceAttrBoolean(val);
  }

  private _required = false;
}
