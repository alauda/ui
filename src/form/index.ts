export * from './common-form';
export * from './form.directive';
export * from './form.module';
export * from './form.types';
export * from './form-item/form-item.component';
export * from './helper-directives';
