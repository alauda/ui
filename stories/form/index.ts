import {
  ButtonModule,
  CheckboxModule,
  FormItemWidth,
  FormModule,
  IconModule,
  InputModule,
  LabelPosition,
  RadioModule,
  SelectModule,
  SwitchModule,
} from '@alauda/ui';
import { boolean, select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/angular';

storiesOf('Form', module)
  .addDecorator(withKnobs)
  .add('form', () => {
    const labelWidth = text('labelWidth', '114px');

    const labelPositionOptions = {
      [LabelPosition.Right]: LabelPosition.Right,
      [LabelPosition.Left]: LabelPosition.Left,
      [LabelPosition.Top]: LabelPosition.Top,
    };
    const contentWidthOptions = {
      [FormItemWidth.Small]: FormItemWidth.Small,
      [FormItemWidth.Medium]: FormItemWidth.Medium,
      [FormItemWidth.Large]: FormItemWidth.Large,
      默认: '',
    };
    const labelPosition = select(
      'labelPosition',
      labelPositionOptions,
      LabelPosition.Right,
    );
    const contentWidth = select('contentWidth', contentWidthOptions, '');

    const inline = boolean('inline', false);

    return {
      moduleMetadata: {
        imports: [
          FormModule,
          InputModule,
          IconModule,
          SelectModule,
          ButtonModule,
          SwitchModule,
          RadioModule,
          CheckboxModule,
        ],
      },
      template: /* HTML */ `
        <form
          #form="ngForm"
          auiForm
          [auiFormLabelWidth]="labelWidth"
          [auiFormLabelPosition]="labelPosition"
          [auiFormInline]="inline"
          [auiFormEmptyAddon]="true"
          style="margin: auto; max-width: 960px;"
        >
          <aui-form-item [width]="contentWidth">
            <label auiFormItemLabel>姓名</label>
            <input
              auiFormItemControl
              name="name"
              #name="ngModel"
              aui-input
              required
              minlength="3"
              [(ngModel)]="value1"
            />
            <div auiFormItemError *ngIf="name.errors?.required">required</div>
            <div auiFormItemError *ngIf="name.errors?.minlength">
              minlength: 3
            </div>
            <div auiFormItemHint>input your name</div>
            <div auiFormItemAddon>
              <aui-icon icon="info_circle_s"></aui-icon>
            </div>
          </aui-form-item>
          <aui-form-item [width]="contentWidth">
            <label auiFormItemLabel>性别</label>
            <aui-select
              auiFormItemControl
              name="sex"
              #sex="ngModel"
              required
              [(ngModel)]="value2"
            >
              <aui-option value="test">test</aui-option>
            </aui-select>
            <div auiFormItemAddon>
              <aui-icon icon="info_circle_s"></aui-icon>
            </div>
          </aui-form-item>
          <aui-form-item>
            <label auiFormItemLabel>年龄</label>
            <aui-number-input
              auiFormItemControl
              name="age"
              #age="ngModel"
              [(ngModel)]="value3"
              required
              min="0"
            ></aui-number-input>
          </aui-form-item>
          <aui-form-item>
            <label auiFormItemLabel>food</label>
            <aui-radio-group name="food" [(ngModel)]="food">
              <aui-radio-button value="apple">apple</aui-radio-button>
              <aui-radio-button value="coffee">coffee</aui-radio-button>
              <aui-radio-button value="tea" disabled="true">
                tea
              </aui-radio-button>
              <aui-radio-button value="icecream">icecream</aui-radio-button>
            </aui-radio-group>
            <div auiFormItemAddon>
              <aui-icon icon="info_circle_s"></aui-icon>
            </div>
          </aui-form-item>
          <aui-form-item>
            <label auiFormItemLabel>这是一个超长的标签</label>
            <aui-radio-group name="food1" [(ngModel)]="food">
              <aui-radio value="apple">apple</aui-radio>
              <aui-radio value="coffee">coffee</aui-radio>
              <aui-radio value="tea" disabled="true">tea</aui-radio>
              <aui-radio value="icecream">icecream</aui-radio>
            </aui-radio-group>
            <div auiFormItemAddon>
              <aui-icon icon="info_circle_s"></aui-icon>
            </div>
          </aui-form-item>
          <aui-form-item>
            <label auiFormItemLabel>long long english label</label>
            <aui-checkbox-group
              auiFormItemControl
              required
              name="foods"
              [(ngModel)]="foods"
            >
              <aui-checkbox label="apple">apple</aui-checkbox>
              <aui-checkbox label="coffee">coffee</aui-checkbox>
              <aui-checkbox label="tea" disabled="true">tea</aui-checkbox>
              <aui-checkbox label="icecream">icecream</aui-checkbox>
            </aui-checkbox-group>
            <div auiFormItemAddon>
              <aui-icon icon="info_circle_s"></aui-icon>
            </div>
          </aui-form-item>
          <aui-form-item>
            <label auiFormItemLabel>开关</label>
            <aui-switch
              name="switch"
              auiFormItemControl
              [(ngModel)]="switchValue"
            ></aui-switch>
            <div auiFormItemAddon>
              <aui-icon icon="info_circle_s"></aui-icon>
            </div>
          </aui-form-item>
          <aui-form-item [width]="contentWidth">
            <label auiFormItemLabel>描述</label>
            <textarea auiFormItemControl aui-input></textarea>
          </aui-form-item>
          <div style="display: flex; justify-content: flex-end;">
            <button aui-button="primary" (click)="form.onSubmit(null)">
              submit
            </button>
            <button aui-button (click)="form.reset(); $event.preventDefault()">
              reset
            </button>
          </div>
        </form>
      `,
      props: {
        labelWidth,
        labelPosition,
        inline,
        contentWidth,
      },
    };
  });
