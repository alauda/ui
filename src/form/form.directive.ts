import { Directive, Inject, Input, Optional } from '@angular/core';
import { ControlContainer, FormGroupDirective, NgForm } from '@angular/forms';
import { BehaviorSubject, Observable, distinctUntilChanged } from 'rxjs';

import { LabelPosition } from './form.types';

@Directive({
  selector: 'form[auiForm]',
  exportAs: 'auiForm',
  host: {
    '[class.aui-form]': 'true',
    '[class.aui-form--inline]': 'inline',
    '[class.ng-submitted]': 'ngClassSubmitted',
  },
})
export class FormDirective {
  get ngClassSubmitted() {
    return (this.controlContainer?.formDirective as NgForm | FormGroupDirective)
      .submitted;
  }

  private readonly labelWidth$$ = new BehaviorSubject<string>(null);
  private readonly labelPosition$$ = new BehaviorSubject<LabelPosition>(
    LabelPosition.Right,
  );

  private readonly emptyAddon$$ = new BehaviorSubject<boolean>(false);

  @Input('auiFormLabelWidth')
  set labelWidth(val: string) {
    this.labelWidth$$.next(val);
  }

  @Input('auiFormLabelPosition')
  set labelPosition(val: LabelPosition) {
    this.labelPosition$$.next(val);
  }

  @Input('auiFormEmptyAddon')
  set emptyAddon(val: boolean) {
    this.emptyAddon$$.next(val);
  }

  @Input('auiFormInline')
  inline = false;

  labelWidth$: Observable<string> = this.labelWidth$$
    .asObservable()
    .pipe(distinctUntilChanged());

  labelPosition$: Observable<LabelPosition> = this.labelPosition$$
    .asObservable()
    .pipe(distinctUntilChanged());

  emptyAddon$: Observable<boolean> = this.emptyAddon$$
    .asObservable()
    .pipe(distinctUntilChanged());

  constructor(
    @Inject(ControlContainer)
    @Optional()
    public controlContainer?: ControlContainer,
  ) {}
}
