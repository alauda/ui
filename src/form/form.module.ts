import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { FormItemComponent } from './form-item/form-item.component';
import { FormDirective } from './form.directive';
import {
  FormItemAddonDirective,
  FormItemControlDirective,
  FormItemErrorDirective,
  FormItemHintDirective,
  FormItemLabelDirective,
} from './helper-directives';

@NgModule({
  imports: [
    CommonModule,
    FormDirective,
    FormItemComponent,
    FormItemAddonDirective,
    FormItemErrorDirective,
    FormItemHintDirective,
    FormItemLabelDirective,
    FormItemControlDirective,
  ],
  exports: [
    FormDirective,
    FormItemComponent,
    FormItemAddonDirective,
    FormItemErrorDirective,
    FormItemHintDirective,
    FormItemLabelDirective,
    FormItemControlDirective,
  ],
})
export class FormModule {}

export const FORM_MODULE = [
  FormDirective,
  FormItemComponent,
  FormItemAddonDirective,
  FormItemErrorDirective,
  FormItemHintDirective,
  FormItemLabelDirective,
  FormItemControlDirective,
] as const;
