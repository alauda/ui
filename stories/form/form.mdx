import { Canvas, Meta } from '@storybook/blocks';

import * as basic from './basic.stories';
import * as inline from './inline.stories';
import * as label from './label.stories';

<Meta title="Example/Form" />

# Form

## 基础使用

<Canvas
  of={basic.Basic}
  meta={basic}
></Canvas>

## Form position top

<Canvas
  of={label.Label}
  meta={label}
></Canvas>

## Form inline

<Canvas
  of={inline.Inline}
  meta={inline}
></Canvas>

## aui-form Inputs

| 名称                 | 类型                       | 默认值 | 描述                   |
| -------------------- | -------------------------- | ------ | ---------------------- |
| auiFormLabelWidth    | string                     | -      | label 宽度             |
| auiFormLabelPosition | 'left' \| 'top' \| 'right' | -      | label 位置             |
| auiFormInline        | boolean                    | -      | 表单结构是否为内联结构 |
| auiFormEmptyAddon    | boolean                    | -      | 是否图标占位           |

## aui-form-item Inputs

| 名称  | 类型                           | 默认值 | 描述                       |
| ----- | ------------------------------ | ------ | -------------------------- |
| width | 'large' \| 'medium' \| 'small' | -      | item 的 content 宽度       |
| plain | boolean                        | false  | control 为纯文本时缩小行高 |
