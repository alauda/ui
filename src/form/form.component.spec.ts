import { Component, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule, NgForm } from '@angular/forms';

import { FormModule } from './form.module';

describe('FormComponent', () => {
  let fixture: ComponentFixture<TestComponent>;
  let ins: TestComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [FormsModule, FormModule],
      declarations: [TestComponent],
    });
    fixture = TestBed.createComponent(TestComponent);
    fixture.detectChanges();
    ins = fixture.componentInstance;
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();

    ins.form.onSubmit(null);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});

@Component({
  template: `
    <form
      #form="ngForm"
      auiForm
      auiFormLabelWidth="100px"
      auiFormLabelPosition="left"
      [auiFormInline]="false"
      [auiFormEmptyAddon]="true"
    >
      <aui-form-item>
        <label auiFormItemLabel>label</label>
        <input
          auiFormItemControl
          name="name"
          #name="ngModel"
          aui-input
          required
          minlength="3"
          [(ngModel)]="value1"
        />
        <div
          auiFormItemError
          *ngIf="name.errors?.required"
        >
          required
        </div>
        <div
          auiFormItemError
          *ngIf="name.errors?.minlength"
        >
          minlength: 3
        </div>
        <div auiFormItemHint>input your name</div>
        <div auiFormItemAddon>addon</div>
      </aui-form-item>
    </form>
  `,
})
class TestComponent {
  @ViewChild('form', { static: true })
  form: NgForm;
}
