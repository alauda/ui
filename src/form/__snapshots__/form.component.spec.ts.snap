// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormComponent should match snapshot 1`] = `
<ng-component
  form={[Function NgForm]}
>
  <form
    auiform=""
    auiformlabelposition="left"
    auiformlabelwidth="100px"
    class="ng-untouched ng-pristine ng-valid aui-form"
    novalidate=""
  >
    <aui-form-item>
      <div
        class="aui-form-item aui-form-item--left"
      >
        <div
          class="aui-form-item__label-wrapper aui-form-item__label-wrapper--required aui-form-item__label-wrapper--left hasLabel"
          style="width: 100px;"
        >
          <label
            auiformitemlabel=""
            class="aui-form-item__label"
          >
            label
          </label>
        </div>
        <div
          class="aui-form-item__container"
        >
          <div
            class="aui-form-item__content"
          >
            <input
              aui-input=""
              auiformitemcontrol=""
              class="ng-untouched ng-pristine ng-valid aui-form-item__control"
              minlength="3"
              name="name"
              required=""
            />
            
            <div
              class="aui-form-item__addon-wrapper"
            >
              <div
                auiformitemaddon=""
                class="aui-form-item__addon"
              >
                addon
              </div>
            </div>
          </div>
          
          
          <div
            class="aui-form-item__hint-wrapper"
          >
            <div
              auiformitemhint=""
              class="aui-form-item__hint"
            >
              input your name
            </div>
          </div>
        </div>
      </div>
    </aui-form-item>
  </form>
</ng-component>
`;

exports[`FormComponent should match snapshot 2`] = `
<ng-component
  form={[Function NgForm]}
>
  <form
    auiform=""
    auiformlabelposition="left"
    auiformlabelwidth="100px"
    class="ng-untouched ng-pristine aui-form ng-invalid ng-submitted"
    novalidate=""
  >
    <aui-form-item>
      <div
        class="aui-form-item aui-form-item--left hasError"
      >
        <div
          class="aui-form-item__label-wrapper aui-form-item__label-wrapper--required aui-form-item__label-wrapper--left hasLabel"
          style="width: 100px;"
        >
          <label
            auiformitemlabel=""
            class="aui-form-item__label"
          >
            label
          </label>
        </div>
        <div
          class="aui-form-item__container"
        >
          <div
            class="aui-form-item__content"
          >
            <input
              aui-input=""
              auiformitemcontrol=""
              class="ng-untouched ng-pristine aui-form-item__control ng-invalid"
              minlength="3"
              name="name"
              required=""
            />
            
            <div
              class="aui-form-item__addon-wrapper"
            >
              <div
                auiformitemaddon=""
                class="aui-form-item__addon"
              >
                addon
              </div>
            </div>
          </div>
          
          
          <div
            class="aui-form-item__hint-wrapper"
          >
            <div
              auiformitemhint=""
              class="aui-form-item__hint"
            >
              input your name
            </div>
          </div>
        </div>
      </div>
    </aui-form-item>
  </form>
</ng-component>
`;
